src/stores/TransitionState.ts

```diff
@@ -0,0 +1,117 @@
+import { RouterState } from './RouterState';
+import { RouterStore, Route } from './RouterStore';
+import { valueEqual } from '@react-force/utils';
+import Debug from 'debug';
+
+const debug = Debug('msr:RouterStore');
+
+export class TransitionState {
+    private transitionState?: RouterState;
+    private readonly fromRoute: Route;
+
+    constructor(
+        private routerStore: RouterStore,
+        private readonly fromState: RouterState
+    ) {
+        // Create an object copy so we can modify hooks during transition.
+        this.fromRoute = Object.assign(
+            {},
+            this.routerStore.getRoute(this.fromState.routeName)
+        );
+    }
+
+    resolve(toState: RouterState): Promise<RouterState> {
+        return this.transition(toState);
+    }
+
+    private async transition(toState: RouterState): Promise<RouterState> {
+        debug('transition from %o to %o)', this.fromState, toState);
+        if (valueEqual(this.transitionState, toState)) {
```

Comment:
What if the loop doesn't involve the final state? E.g. the fromState is A and the toState is B, and A.onExit redirects to C, and C redirects to A.

I think to detect a loop you're going to have to keep track of all the transition states.

There's also the pathological case where a parameter is modified each time to a unique value, so there is never a loop but the recursion is infinite. If we wanted to guard against this we could introduce a maximum number of redirects.

Reply:
You're right, that should be changed. I agree with you that a simple counter will be more flexible. I will change it.